File: apps/api/src/proxy/dispatch_walk.ts
/**
 * The one candidate walk every dispatch surface shares
 * (docs/project-structure.md § Boundaries, docs/providers.md § Routing module).
 *
 * Plans the candidate list, then for each candidate: acquire (decrypt +
 * refresh), fetch with the first-byte timeout and a single 529 retry, ask
 * `routing/feedback` whether the response is a bench-type failure, and
 * either move on or hand the response back. It never shapes the client
 * response and never branches on protocol — the transports in
 * `dispatch.ts` turn its `WalkOutcome` into HTTP or SSE frames.
 */
import { decryptJson } from "../crypto/token_crypto"
import { recordEdgeTimeoutStrike } from "../db/accounts"
import type { CustomProviderRow } from "../db/custom_providers"
import type { Env, ProviderId } from "../env"
import { markBenched } from "../pool/bench"
import type { AcquiredAccount, StoredCredential } from "../pool/acquire"
import { settleLease, type AttemptLease, type PoolExtension } from "../pool/extension"
import { getAdapter } from "../providers"
import { refreshAccountUsageInBackground } from "../providers/usage_refresh"
import type { ProviderAdapter } from "../providers/types"
import { poolCandidates } from "../routing/candidates"
import { candidateFactsList, earliestUnusableUntil } from "../routing/facts"
import {
  EDGE_TIMEOUT_COOLDOWN_MS,
  agentFaultVerdict,
  isEdgeTimeoutStatus,
  penaltyForOutcome,
} from "../routing/feedback"
import { normalizeStrategy, orderCandidates } from "../routing/strategy"
import type { RoutingCandidate } from "../routing/types"

/** Keeps a Worker invocation alive for a deferred `logRequest` past the returned Response — `c.executionCtx.waitUntil` in production, a test double in tests. */
export type WaitUntil = (promise: Promise<unknown>) => void

/** The candidate walk never retries more than this many real upstream calls in one request — same cap as the old `acquireAccount`+exclude loop. */
const MAX_ATTEMPTS = 8

/**
 * The routing module's candidate list, ready for dispatch to walk
 * (docs/providers.md § Routing module). `candidates` and `strategy` let a
 * caller (a group dispatch, via the cross-target flattened list) hand in an
 * already-built, possibly cross-provider list; when omitted, this builds
 * the ordinary single-pool list dispatch has always used — the shape every
 * existing single-provider call site (including tests) still gets.
 */
export type CandidateSource = {
  userId: string
  apiKeyId: string | null
  provider: string
  adapter?: ProviderAdapter
  pinnedAccountId?: string
  /** Pre-built cross-target candidate list (group dispatch) — bypasses the single-pool builder below entirely. */
  candidates?: RoutingCandidate[]
  /** `model_groups.strategy` (candidates given) or `provider_settings.strategy` (single pool); defaults to `ordered`. */
  strategy?: string
  isBuiltin?: boolean
  customProvider?: CustomProviderRow
  /** Composition-time cross-user pool sharing (docs/cloud-edition.md § "Pool extension"); absent for standalone. */
  poolExtension?: PoolExtension
}

type Plan =
  | { kind: "no_account" }
  | { kind: "unavailable"; untilMs: number | null }
  | { kind: "usable"; ordered: RoutingCandidate[] }

async function planCandidates(env: Env, src: CandidateSource, upstreamModel: string): Promise<Plan> {
  const candidates =
    src.candidates ??
    (await poolCandidates(
      env,
      src.userId,
      {
        provider: src.provider,
        upstreamModel,
        isBuiltin: src.isBuiltin ?? true,
        customProvider: src.customProvider,
        adapter: src.adapter ?? getAdapter(src.provider as ProviderId),
        accountId: src.pinnedAccountId ?? null,
      },
      src.poolExtension,
    ))
  if (candidates.length === 0) return { kind: "no_account" }
  const facts = await candidateFactsList(env, src.userId, candidates)
  const ordered = orderCandidates(candidates, facts, {
    apiKeyId: src.apiKeyId,
    strategy: normalizeStrategy(src.strategy),
  })
  const usable = ordered.filter((o) => o.facts.usable).map((o) => o.candidate)
  if (usable.length === 0) return { kind: "unavailable", untilMs: earliestUnusableUntil(facts) }
  return { kind: "usable", ordered: usable }
}

/** Decrypt + `refreshIfNeeded` for one candidate — `null` on an unreadable credential (skipped, never counted as an attempt, mirrors the old `acquireAccount`). */
async function acquireCandidate(env: Env, candidate: RoutingCandidate): Promise<AcquiredAccount | null> {
  try {
    const credential = await decryptJson<StoredCredential>(env.TOKEN_ENCRYPTION_KEY, candidate.account.encrypted_payload)
    let acquired: AcquiredAccount = { row: candidate.account, credential }
    if (candidate.adapter.refreshIfNeeded) acquired = await candidate.adapter.refreshIfNeeded(env, acquired)
    return acquired
  } catch {
    return null
  }
}

/** Per-attempt upstream response-header deadline (docs/api.md "Keepalive and idle timeout"). */
const DEFAULT_FIRST_BYTE_TIMEOUT_MS = 180_000

/** Invalid, absent, or non-positive environment values use the documented default. */
function firstByteTimeoutMs(env: Env): number {
  const parsed = Number(env.UPSTREAM_FIRST_BYTE_TIMEOUT_MS)
  return Number.isFinite(parsed) && parsed > 0 ? Math.floor(parsed) : DEFAULT_FIRST_BYTE_TIMEOUT_MS
}

type TimedUpstreamResponse = { response: Response; timedOut: false } | { response: null; timedOut: true }

/** Wait only for upstream response headers, and always release the timer once fetch resolves or rejects. */
async function fetchUpstreamResponse(
  env: Env,
  attempt: (signal: AbortSignal) => Promise<Response>,
): Promise<TimedUpstreamResponse> {
  const controller = new AbortController()
  let timedOut = false
  const timer = setTimeout(() => {
    timedOut = true
    controller.abort()
  }, firstByteTimeoutMs(env))
  try {
    return { response: await attempt(controller.signal), timedOut: false }
  } catch (error) {
    if (timedOut) return { response: null, timedOut: true }
    throw error
  } finally {
    clearTimeout(timer)
  }
}

function waitForOverloadRetry(): Promise<void> {
  // Jitter keeps concurrent overloaded calls from immediately synchronizing;
  // zero is intentionally allowed in unit tests that stub the retry response.
  return new Promise((resolve) => setTimeout(resolve, 900 + Math.floor(Math.random() * 200)))
}

/** A 529 is transient fleet overload, so retry this account exactly once without benching it. */
async function retry529Once(call: () => Promise<TimedUpstreamResponse>): Promise<TimedUpstreamResponse> {
  const result = await call()
  if (result.timedOut || result.response.status !== 529) return result
  await cancelBody(result.response)
  await waitForOverloadRetry()
  return call()
}

async function cancelBody(res: Response | null): Promise<void> {
  try {
    await res?.body?.cancel()
  } catch {
    /* an abandoned upstream body has nothing left to tell us */
  }
}

/**
 * Bench persistence is feedback, never a reason to abort an in-flight
 * failover walk. Bench writes are scoped to the row's **owner**
 * (`account.user_id`), which is the caller for every own row and the sharer
 * for a borrowed one (docs/cloud-edition.md § "Pool extension") — a shared
 * account that just failed must really get benched, not silently no-op.
 */
async function persistBench(
  env: Env,
  candidate: RoutingCandidate,
  cooldownMs: number,
  upstreamStatus: number,
): Promise<void> {
  try {
    await markBenched(
      env,
      candidate.account.user_id,
      candidate.provider,
      candidate.account.id,
      cooldownMs,
      String(upstreamStatus),
    )
  } catch (error) {
    console.error("Failed to persist account bench", {
      accountId: candidate.account.id,
      provider: candidate.provider,
      error: error instanceof Error ? error.message : String(error),
    })
  }
}

/**
 * Edge-timeout feedback is non-fatal just like bench persistence. Every edge
 * status remains excluded from this walk; only the atomic third strike earns a
 * bench. Returning false therefore covers strikes 1–2 and a failed write.
 */
async function persistEdgeTimeoutStrike(env: Env, candidate: RoutingCandidate): Promise<boolean> {
  try {
    return await recordEdgeTimeoutStrike(env.DB, candidate.account.user_id, candidate.provider, candidate.account.id)
  } catch (error) {
    console.error("Failed to persist edge-timeout strike", {
      accountId: candidate.account.id,
      provider: candidate.provider,
      error: error instanceof Error ? error.message : String(error),
    })
    return false
  }
}

/** Apply feedback and decide whether this pre-stream response must fail over. */
async function shouldFailOverForResponse(
  env: Env,
  candidate: RoutingCandidate,
  response: Response,
): Promise<boolean> {
  // Agent-tunnel fault (docs/cli.md § Failover semantics): infrastructure
  // failure between the DO and the CLI — always the next candidate; only
  // `offline` benches (60s), and reconnect clears that bench.
  const agentFault = agentFaultVerdict(response.headers)
  if (agentFault) {
    if (agentFault.benchMs !== null) {
      try {
        await markBenched(
          env,
          candidate.account.user_id,
          candidate.provider,
          candidate.account.id,
          agentFault.benchMs,
          agentFault.reason,
        )
      } catch (error) {
        console.error("Failed to persist agent-fault bench", {
          accountId: candidate.account.id,
          provider: candidate.provider,
          error: error instanceof Error ? error.message : String(error),
        })
      }
    }
    return true
  }
  if (isEdgeTimeoutStatus(response.status)) {
    if (await persistEdgeTimeoutStrike(env, candidate)) {
      await persistBench(env, candidate, EDGE_TIMEOUT_COOLDOWN_MS, response.status)
    }
    return true
  }
  const penalty = penaltyForOutcome(response.status, response.headers, candidate.account)
  if (!penalty) return false
  await persistBench(env, candidate, penalty.cooldownMs, response.status)
  return true
}

/**
 * Recomputes the earliest bench/limit expiry across exactly the candidates
 * this walk just tried and benched — same "exclude set" the old loop used
 * for its bottom-of-loop `Retry-After`, expressed as a fresh facts read
 * (cheap: bounded by `MAX_ATTEMPTS`, and correct even though a 429's
 * cooldown can vary per candidate, unlike the old flat 300s-for-everything
 * assumption baked into `earliestBenchExpiry`).
 */
export async function recomputeUnavailableUntil(
  env: Env,
  userId: string,
  tried: RoutingCandidate[],
): Promise<number | null> {
  if (tried.length === 0) return null
  const facts = await candidateFactsList(env, userId, tried)
  return earliestUnusableUntil(facts)
}

/** One candidate's upstream call, already bound to the request body/headers; the walk supplies the acquired credential and the first-byte abort signal. */
export type UpstreamCall = (acquired: AcquiredAccount, signal: AbortSignal) => Promise<Response>

export type WalkOpts = {
  source: CandidateSource
  /** Bare upstream model id used to build the single-pool candidate list when `source.candidates` is absent. */
  upstreamModel: string
  /** Picks the adapter method for one candidate; `null` skips a candidate whose adapter lacks the endpoint (never counted as an attempt). */
  callFor: (candidate: RoutingCandidate) => UpstreamCall | null
  waitUntil: WaitUntil
  /** Eager transport: the client already went away — stop before the next acquire/fetch and drop any response already in hand. */
  cancelled?: () => boolean
  /** Filled in as the walk goes so a log written after an early exit (cancel, throw) still names the last candidate tried. */
  progress: WalkProgress
}

/** The attempt record: what the transports need to log, no matter how the walk ended. */
export type WalkProgress = {
  /** Last candidate a real upstream call was made for (`null` before the first attempt). */
  candidate: RoutingCandidate | null
  /** HTTP status of the last upstream response seen, bench-type or not. */
  upstreamStatus: number | null
}

export type WalkOutcome =
  /** Zero candidates for this target — retrying can never help (400 / `no_upstream_account`). */
  | { kind: "no_account" }
  /** Candidates exist but every one is benched or over its usage window right now (503 / `upstream_unavailable`, `Retry-After` from the plan). */
  | { kind: "unavailable"; untilMs: number | null }
  /** `cancelled()` fired mid-walk; any response in hand has been cancelled. Never returned when `cancelled` is absent. */
  | { kind: "cancelled" }
  /** The adapter call rejected (network, adapter bug) — terminal, no further candidates (502 / `upstream_error`). */
  | { kind: "fetch_error"; candidate: RoutingCandidate }
  /**
   * Every attempt was a bench-type failure: either the list ran dry
   * (`lastBenched` = the candidate whose bench response ended it) or the
   * attempt cap hit / the rest was undecryptable (`lastBenched` null). Same
   * synthesized 503 either way; `tried` is the exclude set for `Retry-After`.
   */
  | { kind: "exhausted"; tried: RoutingCandidate[]; lastBenched: RoutingCandidate | null }
  /**
   * A non-bench response — success or a terminal upstream error — for the
   * transport to deliver. `lease` is this attempt's still-open pool-extension
   * lease (docs/cloud-edition.md § "Pool extension"): the transport settles
   * it exactly once, where it decides the attempt's `request_logs` row. Every
   * other outcome above has already settled its own leases `released`.
   */
  | { kind: "response"; candidate: RoutingCandidate; response: Response; lease: AttemptLease | null }

export async function walkCandidates(env: Env, opts: WalkOpts): Promise<WalkOutcome> {
  const cancelled = opts.cancelled ?? (() => false)
  const ext = opts.source.poolExtension
  const plan = await planCandidates(env, opts.source, opts.upstreamModel)
  if (cancelled()) return { kind: "cancelled" }
  if (plan.kind === "no_account") return { kind: "no_account" }
  if (plan.kind === "unavailable") return { kind: "unavailable", untilMs: plan.untilMs }

  let lastResponse: Response | null = null
  let lastCandidate: RoutingCandidate | null = null
  let idx = 0
  let attempts = 0
  let ranOutOfCandidates = false
  for (; attempts < MAX_ATTEMPTS; ) {
    if (cancelled()) {
      await cancelBody(lastResponse)
      return { kind: "cancelled" }
    }
    const candidate = plan.ordered[idx]
    if (!candidate) {
      ranOutOfCandidates = true
      break
    }
    idx++
    const call = opts.callFor(candidate)
    if (!call) continue

    // Pool extension (docs/cloud-edition.md § "Pool extension"): every
    // attempt, own row or shared, is offered for admission immediately
    // before the acquire. A `skip` is exhaustion of that row's governed
    // budget — move on without spending one of MAX_ATTEMPTS on it.
    let lease: AttemptLease | null = null
    if (ext) {
      const reserved = await ext.reserveAttempt(
        env,
        { userId: opts.source.userId, apiKeyId: opts.source.apiKeyId, upstreamModel: candidate.upstreamModel },
        candidate,
      )
      if (reserved && "skip" in reserved) continue
      lease = reserved
    }

    const acquired = await acquireCandidate(env, candidate)
    if (cancelled()) {
      await settleLease(lease, "released")
      await cancelBody(lastResponse)
      return { kind: "cancelled" }
    }
    if (!acquired) {
      await settleLease(lease, "released")
      continue
    }
    attempts++
    opts.progress.candidate = candidate

    let fetched: TimedUpstreamResponse
    try {
      fetched = await retry529Once(() => fetchUpstreamResponse(env, (signal) => call(acquired, signal)))
    } catch {
      // Terminal for this request and logged as `upstream_error` with no
      // output, so the attempt is released, not consumed.
      await settleLease(lease, "released")
      await cancelBody(lastResponse)
      return { kind: "fetch_error", candidate }
    }
    if (fetched.timedOut) {
      await settleLease(lease, "released")
      continue
    }
    const res = fetched.response
    opts.progress.upstreamStatus = res.status
    opts.waitUntil(refreshAccountUsageInBackground(env, candidate.account, candidate.adapter))
    if (cancelled()) {
      await settleLease(lease, "released")
      await cancelBody(lastResponse)
      await cancelBody(res)
      return { kind: "cancelled" }
    }

    if (await shouldFailOverForResponse(env, candidate, res)) {
      // Benched / edge-timed-out / retried: this attempt produced nothing.
      await settleLease(lease, "released")
      await cancelBody(lastResponse)
      lastResponse = res
      lastCandidate = candidate
      continue
    }

    await cancelBody(lastResponse)
    return { kind: "response", candidate, response: res, lease }
  }

  await cancelBody(lastResponse)
  return {
    kind: "exhausted",
    tried: plan.ordered.slice(0, idx),
    lastBenched: ranOutOfCandidates && lastResponse ? lastCandidate : null,
  }
}

File: apps/server/crates/kano-proxy-core/src/proxy/dispatch_walk.rs
//! The candidate walk (docs/providers.md § Routing module,
//! docs/cloud-edition.md § "Pool extension").
//!
//! The one candidate walk every dispatch surface shares. It plans the candidate list, then for
//! each candidate: reserve the attempt with the pool extension, acquire (decrypt + refresh),
//! call the adapter under the first-byte timeout with a single 529 retry, ask
//! [`crate::routing::feedback`] whether the response is a bench-type failure, and either move
//! on or hand the response back. It never shapes the client response and never branches on
//! protocol — the transports in [`crate::proxy::dispatch`] turn a [`WalkOutcome`] into HTTP or
//! SSE frames.

use std::sync::{Arc, Mutex};
use std::time::Duration;

use async_trait::async_trait;
use axum::response::Response;
use rand::Rng;

use crate::app::now_ms;
use crate::db::accounts::record_edge_timeout_strike;
use crate::db::custom_providers::CustomProviderRow;
use crate::pool::acquire::{acquire, AcquiredAccount};
use crate::pool::bench::mark_benched;
use crate::pool::extension::{settle_lease, AttemptLease, LeaseOutcome, ReserveContext, ReserveOutcome};
use crate::providers::types::{AdapterError, CallExtras};
use crate::providers::{DynAdapter, ProviderId};
use crate::routing::candidates::{pool_candidates, PoolTarget};
use crate::routing::facts::{candidate_facts_list, earliest_unusable_until};
use crate::routing::feedback::{
    agent_fault_verdict, is_edge_timeout_status, penalty_for_outcome, EDGE_TIMEOUT_COOLDOWN_MS,
};
use crate::routing::strategy::{normalize_strategy, order_candidates};
use crate::routing::types::{RoutingCandidate, StrategyContext};
use crate::AppState;

/// The candidate walk never makes more than this many real upstream calls in one request.
pub const MAX_ATTEMPTS: usize = 8;

/// `cancelled()` for the eager transport: the client already went away.
pub type CancelSignal = Arc<dyn Fn() -> bool + Send + Sync>;

/// The routing module's candidate list, ready for dispatch to walk. `candidates` and
/// `strategy` let a caller (a group dispatch, via the cross-target flattened list) hand in an
/// already-built, possibly cross-provider list; when absent, this builds the ordinary
/// single-pool list dispatch has always used.
///
/// Deviation from the TypeScript: the pool extension is not a field here — it comes from
/// `cx.pool_extension()`, the composition seam an edition already installed on the `AppState`.
#[derive(Clone, Default)]
pub struct CandidateSource {
    pub user_id: String,
    pub api_key_id: Option<String>,
    pub provider: String,
    /// Pre-resolved adapter for custom providers; defaults to the builtin registry.
    pub adapter: Option<DynAdapter>,
    /// Model-group account pinning: restrict acquire/failover to exactly this row.
    pub pinned_account_id: Option<String>,
    /// Pre-built cross-target candidate list (group dispatch) — bypasses the single-pool
    /// builder entirely.
    pub candidates: Option<Vec<RoutingCandidate>>,
    /// `model_groups.strategy` (candidates given) or `provider_settings.strategy` (single
    /// pool); defaults to `ordered`.
    pub strategy: Option<String>,
    pub is_builtin: Option<bool>,
    pub custom_provider: Option<CustomProviderRow>,
}

enum Plan {
    NoAccount,
    Unavailable { until_ms: Option<i64> },
    Usable { ordered: Vec<RoutingCandidate> },
}

async fn plan_candidates(cx: &AppState, src: &CandidateSource, upstream_model: &str) -> Result<Plan, sqlx::Error> {
    let candidates = match src.candidates.clone() {
        Some(candidates) => candidates,
        None => {
            let is_builtin = src.is_builtin.unwrap_or(true);
            let adapter = match src.adapter.clone() {
                Some(adapter) => adapter,
                None => match ProviderId::parse(&src.provider) {
                    Some(id) => crate::providers::registry::get_adapter(id),
                    // A non-builtin provider with no adapter cannot be called at all; an empty
                    // pool is the honest answer (`no_upstream_account`).
                    None => return Ok(Plan::NoAccount),
                },
            };
            pool_candidates(
                cx,
                &src.user_id,
                &PoolTarget {
                    provider: src.provider.clone(),
                    upstream_model: upstream_model.to_string(),
                    is_builtin,
                    custom_provider: src.custom_provider.clone(),
                    adapter,
                    account_id: src.pinned_account_id.clone(),
                },
            )
            .await?
        }
    };
    if candidates.is_empty() {
        return Ok(Plan::NoAccount);
    }
    let now = now_ms();
    let facts = candidate_facts_list(&candidates, now);
    let ordered = order_candidates(
        candidates,
        &facts,
        &StrategyContext {
            api_key_id: src.api_key_id.clone(),
            strategy: normalize_strategy(src.strategy.as_deref()),
        },
    );
    let usable: Vec<RoutingCandidate> =
        ordered.into_iter().filter(|o| o.facts.usable).map(|o| o.candidate).collect();
    if usable.is_empty() {
        return Ok(Plan::Unavailable { until_ms: earliest_unusable_until(&facts) });
    }
    Ok(Plan::Usable { ordered: usable })
}

/// Decrypt + `refresh_if_needed` for one candidate — `None` on an unreadable credential
/// (skipped, never counted as an attempt).
async fn acquire_candidate(cx: &AppState, candidate: &RoutingCandidate) -> Option<AcquiredAccount> {
    let acquired = match acquire(cx, candidate.account.clone()) {
        Ok(acquired) => acquired,
        Err(error) => {
            tracing::debug!(account_id = %candidate.account.id, %error, "candidate credential is unreadable");
            return None;
        }
    };
    match candidate.adapter.refresh_if_needed(cx, acquired).await {
        Ok(acquired) => Some(acquired),
        Err(error) => {
            tracing::debug!(account_id = %candidate.account.id, %error, "candidate refresh failed");
            None
        }
    }
}

/// Per-attempt upstream response-header deadline (docs/api.md § Keepalive and idle timeout).
fn first_byte_timeout(cx: &AppState) -> Duration {
    Duration::from_millis(cx.config().upstream_first_byte_timeout_ms)
}

enum Fetched {
    Response(Response),
    /// No response headers inside the deadline — fail over without benching.
    TimedOut,
    /// The adapter itself failed (network, adapter bug) — terminal for this request.
    Failed,
}

/// One candidate's upstream call, already bound to the request body/headers; the walk supplies
/// the acquired credential and the per-attempt deadline.
#[async_trait]
pub trait CandidateCaller: Send + Sync {
    /// `false` skips a candidate whose adapter lacks the endpoint (never counted as an attempt).
    fn supports(&self, candidate: &RoutingCandidate) -> bool;

    async fn call(
        &self,
        cx: &AppState,
        candidate: &RoutingCandidate,
        acquired: &AcquiredAccount,
        extras: &CallExtras,
    ) -> Result<Response, AdapterError>;
}

/// Waits only for the adapter to produce response headers. The deadline also reaches the
/// adapter through `CallExtras::first_byte_timeout`, so the transport aborts the request
/// itself; the outer timeout covers an adapter that never returns at all — together they are
/// the `AbortController` the TypeScript walk wrapped every attempt in.
async fn fetch_upstream_response(
    cx: &AppState,
    caller: &dyn CandidateCaller,
    candidate: &RoutingCandidate,
    acquired: &AcquiredAccount,
    extras: &CallExtras,
) -> Fetched {
    match tokio::time::timeout(first_byte_timeout(cx), caller.call(cx, candidate, acquired, extras)).await {
        Err(_) => Fetched::TimedOut,
        Ok(Ok(response)) => Fetched::Response(response),
        Ok(Err(AdapterError::Transport(crate::upstream::TransportError::Timeout))) => Fetched::TimedOut,
        Ok(Err(error)) => {
            tracing::debug!(provider = %candidate.provider, %error, "upstream call failed");
            Fetched::Failed
        }
    }
}

/// Jitter keeps concurrent overloaded calls from immediately synchronizing.
async fn wait_for_overload_retry() {
    let jitter = rand::thread_rng().gen_range(0..200);
    tokio::time::sleep(Duration::from_millis(900 + jitter)).await;
}

/// A 529 is transient fleet overload, so retry this account exactly once without benching it.
async fn retry_529_once(
    cx: &AppState,
    caller: &dyn CandidateCaller,
    candidate: &RoutingCandidate,
    acquired: &AcquiredAccount,
    extras: &CallExtras,
) -> Fetched {
    let result = fetch_upstream_response(cx, caller, candidate, acquired, extras).await;
    match &result {
        Fetched::Response(res) if res.status().as_u16() == 529 => {}
        _ => return result,
    }
    drop(result);
    wait_for_overload_retry().await;
    fetch_upstream_response(cx, caller, candidate, acquired, extras).await
}

/// Bench persistence is feedback, never a reason to abort an in-flight failover walk. Bench
/// writes are scoped to the row's **owner** (`account.user_id`), which is the caller for every
/// own row and the sharer for a borrowed one — a shared account that just failed must really
/// get benched, not silently no-op.
async fn persist_bench(cx: &AppState, candidate: &RoutingCandidate, cooldown_ms: i64, reason: &str) {
    if let Err(error) = mark_benched(
        cx.pool(),
        &candidate.account.user_id,
        &candidate.provider,
        &candidate.account.id,
        Some(cooldown_ms),
        Some(reason),
        now_ms(),
    )
    .await
    {
        tracing::error!(
            account_id = %candidate.account.id,
            provider = %candidate.provider,
            %error,
            "Failed to persist account bench"
        );
    }
}

/// Edge-timeout feedback is non-fatal just like bench persistence. Every edge status remains
/// excluded from this walk; only the atomic third strike earns a bench. `false` therefore
/// covers strikes 1–2 and a failed write.
async fn persist_edge_timeout_strike(cx: &AppState, candidate: &RoutingCandidate) -> bool {
    match record_edge_timeout_strike(
        cx.pool(),
        &candidate.account.user_id,
        &candidate.provider,
        &candidate.account.id,
        now_ms(),
    )
    .await
    {
        Ok(third) => third,
        Err(error) => {
            tracing::error!(
                account_id = %candidate.account.id,
                provider = %candidate.provider,
                %error,
                "Failed to persist edge-timeout strike"
            );
            false
        }
    }
}

/// Apply feedback and decide whether this pre-stream response must fail over.
/// `status` and `headers` rather than the whole response: an `axum::Response` is not `Sync`,
/// and this runs inside the eager transport's `Send` producer future.
async fn should_fail_over_for_response(
    cx: &AppState,
    candidate: &RoutingCandidate,
    status: u16,
    headers: &http::HeaderMap,
) -> bool {
    // Agent-tunnel fault (docs/cli.md § Failover semantics): infrastructure failure between the
    // tunnel and the CLI — always the next candidate; only `offline` benches (60s), and
    // reconnect clears that bench.
    if let Some(fault) = agent_fault_verdict(headers) {
        if let Some(bench_ms) = fault.bench_ms {
            persist_bench(cx, candidate, bench_ms, &fault.reason).await;
        }
        return true;
    }
    if is_edge_timeout_status(status) {
        if persist_edge_timeout_strike(cx, candidate).await {
            persist_bench(cx, candidate, EDGE_TIMEOUT_COOLDOWN_MS, &status.to_string()).await;
        }
        return true;
    }
    let Some(penalty) = penalty_for_outcome(status, headers, &candidate.account, now_ms()) else {
        return false;
    };
    persist_bench(cx, candidate, penalty.cooldown_ms, &status.to_string()).await;
    true
}

/// Recomputes the earliest bench/limit expiry across exactly the candidates this walk just
/// tried and benched — the exclude set behind the bottom-of-loop `Retry-After`, expressed as a
/// fresh facts read (cheap: bounded by [`MAX_ATTEMPTS`], and correct even though a 429's
/// cooldown can vary per candidate).
pub async fn recompute_unavailable_until(cx: &AppState, tried: &[RoutingCandidate]) -> Option<i64> {
    if tried.is_empty() {
        return None;
    }
    let now = now_ms();
    let mut earliest: Option<i64> = None;
    for candidate in tried {
        // The rows were mutated by this walk's bench writes, so they are re-read rather than
        // re-using the copies the plan captured.
        let row = match crate::db::accounts::get_account(cx.pool(), &candidate.account.user_id, &candidate.account.id)
            .await
        {
            Ok(Some(row)) => row,
            Ok(None) => continue,
            Err(error) => {
                tracing::error!(%error, "Failed to re-read a tried candidate for Retry-After");
                continue;
            }
        };
        if let Some(until) = crate::routing::facts::facts_for_row(&row, now).unusable_until {
            earliest = Some(earliest.map_or(until, |e: i64| e.min(until)));
        }
    }
    earliest
}

/// The attempt record: what the transports need to log, no matter how the walk ended.
#[derive(Default, Clone)]
pub struct WalkProgress {
    /// Last candidate a real upstream call was made for (`None` before the first attempt).
    pub candidate: Option<RoutingCandidate>,
    /// HTTP status of the last upstream response seen, bench-type or not.
    pub upstream_status: Option<i32>,
}

pub struct WalkOpts<'a> {
    pub source: &'a CandidateSource,
    /// Bare upstream model id used to build the single-pool candidate list when
    /// `source.candidates` is absent.
    pub upstream_model: &'a str,
    pub caller: &'a dyn CandidateCaller,
    /// Eager transport: the client already went away — stop before the next acquire/fetch.
    pub cancelled: Option<CancelSignal>,
    /// Filled in as the walk goes so a log written after an early exit (cancel, throw) still
    /// names the last candidate tried.
    pub progress: Arc<Mutex<WalkProgress>>,
}

pub enum WalkOutcome {
    /// Zero candidates for this target — retrying can never help (400 / `no_upstream_account`).
    NoAccount,
    /// Candidates exist but every one is benched or over its usage window right now
    /// (503 / `upstream_unavailable`, `Retry-After` from the plan).
    Unavailable { until_ms: Option<i64> },
    /// `cancelled()` fired mid-walk; any response in hand has been dropped. Never returned when
    /// `cancelled` is absent.
    Cancelled,
    /// The adapter call failed (network, adapter bug) — terminal, no further candidates
    /// (502 / `upstream_error`).
    FetchError { candidate: RoutingCandidate },
    /// Every attempt was a bench-type failure: either the list ran dry (`last_benched` = the
    /// candidate whose bench response ended it) or the attempt cap hit / the rest was
    /// undecryptable (`last_benched` `None`). Same synthesized 503 either way; `tried` is the
    /// exclude set for `Retry-After`.
    Exhausted { tried: Vec<RoutingCandidate>, last_benched: Option<RoutingCandidate> },
    /// A non-bench response — success or a terminal upstream error — for the transport to
    /// deliver. `lease` is this attempt's still-open pool-extension lease: the transport settles
    /// it exactly once, where it decides the attempt's `request_logs` row. Every other outcome
    /// has already settled its own leases `Released`.
    Response { candidate: RoutingCandidate, response: Response, lease: Option<Box<dyn AttemptLease>> },
}

pub async fn walk_candidates(cx: &AppState, opts: WalkOpts<'_>) -> WalkOutcome {
    let cancelled = || opts.cancelled.as_ref().is_some_and(|f| f());
    let plan = match plan_candidates(cx, opts.source, opts.upstream_model).await {
        Ok(plan) => plan,
        Err(error) => {
            tracing::error!(%error, "Failed to plan routing candidates");
            return WalkOutcome::NoAccount;
        }
    };
    if cancelled() {
        return WalkOutcome::Cancelled;
    }
    let ordered = match plan {
        Plan::NoAccount => return WalkOutcome::NoAccount,
        Plan::Unavailable { until_ms } => return WalkOutcome::Unavailable { until_ms },
        Plan::Usable { ordered } => ordered,
    };

    let extras =
        CallExtras { api_key_id: opts.source.api_key_id.clone(), first_byte_timeout: Some(first_byte_timeout(cx)) };
    let mut last_candidate: Option<RoutingCandidate> = None;
    let mut saw_bench_response = false;
    let mut idx = 0usize;
    let mut attempts = 0usize;
    let mut ran_out_of_candidates = false;
    while attempts < MAX_ATTEMPTS {
        if cancelled() {
            return WalkOutcome::Cancelled;
        }
        let Some(candidate) = ordered.get(idx).cloned() else {
            ran_out_of_candidates = true;
            break;
        };
        idx += 1;
        if !opts.caller.supports(&candidate) {
            continue;
        }

        // Pool extension (docs/cloud-edition.md § "Pool extension"): every attempt, own row or
        // shared, is offered for admission immediately before the acquire. A `Skip` is
        // exhaustion of that row's governed budget — move on without spending one of
        // MAX_ATTEMPTS on it.
        let mut lease: Option<Box<dyn AttemptLease>> = None;
        if let Some(ext) = cx.pool_extension() {
            let reserved = ext
                .reserve_attempt(
                    cx,
                    ReserveContext {
                        user_id: &opts.source.user_id,
                        api_key_id: opts.source.api_key_id.as_deref(),
                        upstream_model: &candidate.upstream_model,
                    },
                    &candidate,
                )
                .await;
            match reserved {
                Ok(ReserveOutcome::Skip) => continue,
                Ok(ReserveOutcome::Admitted(l)) => lease = Some(l),
                Ok(ReserveOutcome::Ungoverned) => {}
                Err(error) => {
                    tracing::error!(%error, "Pool extension reserveAttempt failed");
                    continue;
                }
            }
        }

        let acquired = acquire_candidate(cx, &candidate).await;
        if cancelled() {
            settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
            return WalkOutcome::Cancelled;
        }
        let Some(acquired) = acquired else {
            settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
            continue;
        };
        attempts += 1;
        opts.progress.lock().unwrap_or_else(|e| e.into_inner()).candidate = Some(candidate.clone());

        let fetched = retry_529_once(cx, opts.caller, &candidate, &acquired, &extras).await;
        let response = match fetched {
            Fetched::Failed => {
                // Terminal for this request and logged as `upstream_error` with no output, so
                // the attempt is released, not consumed.
                settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
                return WalkOutcome::FetchError { candidate };
            }
            Fetched::TimedOut => {
                settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
                continue;
            }
            Fetched::Response(response) => response,
        };
        opts.progress.lock().unwrap_or_else(|e| e.into_inner()).upstream_status =
            Some(response.status().as_u16() as i32);
        crate::providers::usage_refresh::spawn_account_usage_refresh(
            cx,
            candidate.account.clone(),
            candidate.adapter.clone(),
        );
        if cancelled() {
            settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
            return WalkOutcome::Cancelled;
        }

        let status = response.status().as_u16();
        let headers = response.headers().clone();
        if should_fail_over_for_response(cx, &candidate, status, &headers).await {
            // Benched / edge-timed-out / retried: this attempt produced nothing.
            settle_lease(lease.as_deref(), LeaseOutcome::Released).await;
            drop(response);
            saw_bench_response = true;
            last_candidate = Some(candidate);
            continue;
        }

        return WalkOutcome::Response { candidate, response, lease };
    }

    WalkOutcome::Exhausted {
        tried: ordered.into_iter().take(idx).collect(),
        last_benched: if ran_out_of_candidates && saw_bench_response { last_candidate } else { None },
    }
}
